### PR TITLE
[#12] Add CouchDB implementation for id store

### LINKED ISSUE #12: Add CouchDB implementation for id store

### PR DESCRIPTION
**Pull request by fqutishat**
Signed-off-by: Firas Qutishat <firas.qutishat@securekey.com>

closes #12

### CODE REVIEW

**Comment by fqutishat** [comment]
verify-unitest

File: pkg/idstore/store_impl.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+func newStore(couchInstance *couchdb.CouchInstance, dbName string) (idstore.IDStore, error) {
+	db, err := couchdb.NewCouchDatabase(couchInstance, dbName)
+	if err != nil {
+		return nil, errors.WithMessage(err, "create new couchdb database failed %s")
+	}
+
+	dbExists, err := db.ExistsWithRetry()
+	if err != nil {
+		return nil, errors.WithMessage(err, "check couchdb exist failed %s")
```
Comment (Baha-sk) [inline pkg/idstore/store_impl.go]: replace:
return nil, errors.WithMessage(err, "check couchdb exist failed %s")
with 
return nil, errors.WithMessage(err, **fmt.Sprintf("check if couchdb '%s' exists failed", db.DBName)**)

File: pkg/idstore/store_impl.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+		logger.Error(err.Error())
+		return nil
+	}
+	return s
+}
+
+func newStore(couchInstance *couchdb.CouchInstance, dbName string) (idstore.IDStore, error) {
+	db, err := couchdb.NewCouchDatabase(couchInstance, dbName)
+	if err != nil {
+		return nil, errors.WithMessage(err, "create new couchdb database failed %s")
```
Comment (Baha-sk) [inline pkg/idstore/store_impl.go]: replace:
return nil, errors.WithMessage(err, "create new couchdb database failed %s")
with
return nil, errors.WithMessage(err, **fmt.Sprintf("create new couchdb database called '%s' failed" , db.DBName)**)

File: pkg/idstore/store_impl.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+	if err != nil {
+		return nil, errors.WithMessage(err, "check couchdb exist failed %s")
+	}
+	if !dbExists {
+		return nil, errors.New(fmt.Sprintf("DB not found: [%s]", db.DBName))
+	}
+
+	indexExists, err := db.IndexDesignDocExistsWithRetry(inventoryTypeIndexDoc)
+	if err != nil {
+		return nil, errors.WithMessage(err, "check index exist failed %s")
```
Comment (Baha-sk) [inline pkg/idstore/store_impl.go]: replace:
return nil, errors.WithMessage(err, "check index exist failed %s")
with
return nil, errors.WithMessage(err, **fmt.Sprintf("check index exists for '%s' database failed", db.DBName)**)

File: pkg/idstore/store_impl.go (hunk 1 above)
Comment (troyronda) [reply pkg/idstore/store_impl.go]: Should actually be WithMessagef assuming errors package is up-to-date

File: pkg/idstore/store_impl.go (hunk 2 above)
Comment (troyronda) [reply pkg/idstore/store_impl.go]: Should actually be WithMessagef assuming errors package is up-to-date

File: pkg/idstore/store_impl.go (hunk 3 above)
Comment (troyronda) [reply pkg/idstore/store_impl.go]: Should actually be WithMessagef assuming errors package is up-to-date

File: pkg/idstore/store_impl.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+	}
+
+	s := Store{db, ""}
+	return &s, nil
+}
+
+func newCommitterStore(couchInstance *couchdb.CouchInstance, dbName string) idstore.IDStore {
+	db, err := couchdb.CreateCouchDatabase(couchInstance, dbName)
+	if err != nil {
+		logger.Error("create new couchdb database failed %s", err.Error())
```
Comment (Baha-sk) [inline pkg/idstore/store_impl.go]: replace 'logger.Error(' with 'logger.Error**f**('

File: pkg/idstore/store_impl.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+		return nil
+	}
+
+	err = createIndices(db)
+	if err != nil {
+		logger.Error("create couchdb index failed %s", err.Error())
```
Comment (Baha-sk) [inline pkg/idstore/store_impl.go]: same as above

File: pkg/idstore/store_impl.go (hunk 1 above)
Comment (fqutishat) [reply pkg/idstore/store_impl.go]: done

File: pkg/idstore/store_impl.go (hunk 3 above)
Comment (fqutishat) [reply pkg/idstore/store_impl.go]: done

File: pkg/idstore/store_impl.go (hunk 2 above)
Comment (fqutishat) [reply pkg/idstore/store_impl.go]: done

File: pkg/idstore/couchdoc_conv.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+func couchValueToJSON(value []byte) (jsonValue, error) {
+	// create a generic map unmarshal the json
+	jsonResult := make(map[string]interface{})
+	decoder := json.NewDecoder(bytes.NewBuffer(value))
+	decoder.UseNumber()
+
+	err := decoder.Decode(&jsonResult)
+	if err != nil {
+		return nil, errors.Wrapf(err, "result from DB is not JSON encoded")
```
Comment (Baha-sk) [inline pkg/idstore/couchdoc_conv.go]: repleace Wrapf with Wrap

File: pkg/idstore/store_impl.go [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+func newStore(couchInstance *couchdb.CouchInstance, dbName string) (idstore.IDStore, error) {
+	db, err := couchdb.NewCouchDatabase(couchInstance, dbName)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "create new couchdb database called [%s] failed", dbName)
+	}
+
+	dbExists, err := db.ExistsWithRetry()
+	if err != nil {
+		return nil, errors.WithMessagef(err, "check couchdb [%s] exist failed %v", dbName)
```
Comment (Baha-sk) [inline pkg/idstore/store_impl.go]: remove **%v** at the end of the the message

File: pkg/idstore/store_impl.go [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+
+	s := Store{db, ""}
+
+	return &s
+}
+
+func createIndices(db *couchdb.CouchDatabase) error {
+	err := db.CreateNewIndexWithRetry(inventoryTypeIndexDef, inventoryTypeIndexDoc)
+	if err != nil {
+		return errors.WithMessage(err, "creation of inventory metadata index failed")
```
Comment (Baha-sk) [inline pkg/idstore/store_impl.go]: maybe change to:
return errors.WithMessage**f(err, "creation of inventory metadata index failed for [%s]", db.DBName)**

?

Comment (fqutishat) [reply pkg/idstore/store_impl.go]: done

Comment (fqutishat) [reply pkg/idstore/store_impl.go]: done

File: pkg/idstore/store_impl.go (hunk 7 above)
Comment (fqutishat) [reply pkg/idstore/store_impl.go]: done

File: pkg/idstore/couchdoc_conv.go (hunk 6 above)
Comment (fqutishat) [reply pkg/idstore/couchdoc_conv.go]: done

File: pkg/idstore/store_impl.go (hunk 5 above)
Comment (fqutishat) [reply pkg/idstore/store_impl.go]: done

File: pkg/idstore/store_impl.go (hunk 4 above)
Comment (fqutishat) [reply pkg/idstore/store_impl.go]: done

**Pull request merged**